src/pose.rs

```diff
@@ -0,0 +1,249 @@
+    let origin = pose.transform_point(&Point3::new(0.0, 0.0, 0.0));
+    let x_axis = pose.transform_point(&Point3::new(length, 0.0, 0.0));
+    let y_axis = pose.transform_point(&Point3::new(0.0, length, 0.0));
+    let z_axis = pose.transform_point(&Point3::new(0.0, 0.0, length));
+    lines.push(Line {
+        x1: origin.x,
+        y1: origin.y,
+        x2: x_axis.x,
+        y2: x_axis.y,
+        color: style::Color::Red,
+    });
+    lines.push(Line {
+        x1: origin.x,
+        y1: origin.y,
+        x2: y_axis.x,
+        y2: y_axis.y,
+        color: style::Color::Green,
+    });
+    lines.push(Line {
+        x1: origin.x,
+        y1: origin.y,
+        x2: z_axis.x,
+        y2: z_axis.y,
+        color: style::Color::Blue,
+    });
+    lines
+}
+
+fn poses_to_lines(poses: &Vec<Isometry3<f64>>, color: &Color) -> Vec<Line> {
+    poses
+        .windows(2)
+        .map(|w| {
+            let p0 = w[0].transform_point(&Point3::new(0.0, 0.0, 0.0));
+            let p1 = w[1].transform_point(&Point3::new(0.0, 0.0, 0.0));
+            Line {
+                x1: p0.x,
+                y1: p0.y,
+                x2: p1.x,
+                y2: p1.y,
+                color: style::Color::Rgb(color.r, color.g, color.b),
+            }
+        })
+        .collect()
+}
+
+pub struct PoseStampedListener {
+    config: PoseListenerConfig,
+    pose: Arc<RwLock<Option<Isometry3<f64>>>>,
+    _subscriber: rosrust::Subscriber,
+}
+
+impl PoseStampedListener {
+    pub fn new(config: PoseListenerConfig) -> PoseStampedListener {
+        let pose = Arc::new(RwLock::new(None));
+        let cb_pose = pose.clone();
+        let sub = rosrust::subscribe(
+            &config.topic,
+            2,
+            move |pose_msg: rosrust_msg::geometry_msgs::PoseStamped| {
+                let pose_iso = ros_pose_to_isometry(&pose_msg.pose);
+                *cb_pose.write().unwrap() = Some(pose_iso);
+            },
+        )
+        .unwrap();
+
+        PoseStampedListener {
+            config: config,
+            pose: pose,
+            _subscriber: sub,
+        }
+    }
+
+    pub fn get_lines(&self) -> Vec<Line> {
+        match *self.pose.read().unwrap() {
+            Some(p) => match self.config.style.as_str() {
+                "arrow" => pose_to_arrow(&p, self.config.length, &self.config.color),
+                "axes" => pose_to_axes(&p, self.config.length),
+                _ => Vec::new(),
+            },
+            None => Vec::new(),
+        }
+    }
+}
+
+pub struct PoseArrayListener {
+    config: PoseListenerConfig,
+    poses: Arc<RwLock<Vec<Isometry3<f64>>>>,
+    _subscriber: rosrust::Subscriber,
+}
+
+impl PoseArrayListener {
+    pub fn new(config: PoseListenerConfig) -> PoseArrayListener {
+        let poses = Arc::new(RwLock::new(Vec::<Isometry3<f64>>::new()));
+        let cb_poses = poses.clone();
+        let sub = rosrust::subscribe(
+            &config.topic,
+            2,
```

Comment:
we are inconsistent here, why 2?

Reply:
I copied from your listeners. Any suggestion on the size? Should I put 1?